lib/httpserver/httpserver_test.go

```diff
@@ -34,3 +36,175 @@ func TestGetQuotedRemoteAddr(t *testing.T) {
 	f("1.2.3.4", "foo.bar", `"1.2.3.4, X-Forwarded-For: foo.bar"`)
 	f("1.2\n\"3.4", "foo\nb\"ar", `"1.2\n\"3.4, X-Forwarded-For: foo\nb\"ar"`)
 }
+
+func TestBasicAuthMetrics(t *testing.T) {
+	origUsername := *httpAuthUsername
+	origPasswd := *httpAuthPassword
+	testsForConfiguredAuth := []struct {
+		username     string
+		passwd       string
+		expectedCode int
+	}{
+		{
+			username:     "test",
+			passwd:       "pass",
+			expectedCode: 200,
+		},
+		{
+			username:     "test",
+			passwd:       "wrong",
+			expectedCode: 401,
+		},
+		{
+			username:     "wrongUser",
+			passwd:       "pass",
+			expectedCode: 401,
+		},
+		{
+			username:     "wrongUser",
+			passwd:       "wrongpass",
+			expectedCode: 401,
+		},
+	}
+
+	testsWhenAuthIsNotConfigured := []struct {
+		username     string
+		passwd       string
+		expectedCode int
+	}{
+		{
+			username:     "test",
+			passwd:       "pass",
+			expectedCode: 200,
+		},
+		{
+			username:     "test",
+			passwd:       "wrong",
+			expectedCode: 200,
+		},
+		{
+			username:     "wrongUser",
+			passwd:       "pass",
+			expectedCode: 200,
+		},
+		{
+			username:     "wrongUser",
+			passwd:       "wrongpass",
+			expectedCode: 200,
+		},
+	}
+	*httpAuthUsername = "test"
+	*httpAuthPassword = "pass"
+	for _, tt := range testsForConfiguredAuth {
+		req := httptest.NewRequest(http.MethodGet, "/metrics", nil)
+		req.SetBasicAuth(tt.username, tt.passwd)
+
+		w := httptest.NewRecorder()
+		CheckBasicAuth(w, req)
+
+		res := w.Result()
+		defer res.Body.Close()
+		if tt.expectedCode != res.StatusCode {
+			t.Fatalf("Unexpected status code: %d, Expected code is: %d\n", res.StatusCode, tt.expectedCode)
+		}
+	}
+
+	*httpAuthUsername = ""
+	*httpAuthPassword = ""
+	for _, tt := range testsWhenAuthIsNotConfigured {
+		req := httptest.NewRequest(http.MethodGet, "/metrics", nil)
+		req.SetBasicAuth(tt.username, tt.passwd)
+
+		w := httptest.NewRecorder()
+		CheckBasicAuth(w, req)
+
+		res := w.Result()
+		defer res.Body.Close()
+		if tt.expectedCode != res.StatusCode {
+			t.Fatalf("Unexpected status code: %d, Expected code is: %d\n", res.StatusCode, tt.expectedCode)
+		}
+	}
+
+	*httpAuthPassword = origPasswd
+	*httpAuthUsername = origUsername
+
+}
```

Comment:
nit: it could be more readable and compact in the following form:
```
func TestBasicAuthMetrics(t *testing.T) {
	origUsername := *httpAuthUsername
	origPasswd := *httpAuthPassword
	defer func() {
		*httpAuthPassword = origPasswd
		*httpAuthUsername = origUsername
	}()

	f := func(user, pass string, expCode int) {
		t.Helper()
		req := httptest.NewRequest(http.MethodGet, "/metrics", nil)
		req.SetBasicAuth(user, pass)

		w := httptest.NewRecorder()
		CheckBasicAuth(w, req)

		res := w.Result()
		_ = res.Body.Close()
		if expCode != res.StatusCode {
			t.Fatalf("wanted status code: %d, got: %d\n", res.StatusCode, expCode)
		}
	}

	*httpAuthUsername = "test"
	*httpAuthPassword = "pass"
	f("test", "pass", 200)
	f("test", "wrong", 401)
	f("wrong", "pass", 401)
	f("wrong", "wrong", 401)

	*httpAuthUsername = ""
	*httpAuthPassword = ""
	f("test", "pass", 200)
	f("test", "wrong", 200)
	f("wrong", "pass", 200)
	f("wrong", "wrong", 200)
}
```